### DISCUSSION TITLE
next routes should get a unique instance key between similar routes

### DISCUSSION THREAD

**Discussion by CaptainN** [vercel/next.js#32751] [Ideas]
### What version of Next.js are you using?

10.2.0

### What version of Node.js are you using?

14

### What browser are you using?

Any

### What operating system are you using?

macOS

### How are you deploying your application?

Various, most with next start, some with export

### Describe the Bug

When changing between two similar routes - for example, two routes that have the same handler `/pages/[slug].js` the mounted component from the first route is retained, and reused for the next set of data. If the page is using derived data (the props are used to set initial state in `useState` or `useReducer` - the data can become stale. A workaround can be employed, but it involves a lot of complexity through the use of either `useMemo` or `useEffect` with properly configured `deps` arrays.

### Expected Behavior

When the slug changes, it should force a full react lifecycle. This can be done by changing the `key` of the mounted component. For example, if the `/pages/[slug].js` route uses `/pages/about` and it links to `/pages/pool-party` if those values are used for the `key` on the component instance, it'll cause a full lifecycle. Changing this key is unavailable to the end user (to my knowledge - although I supposed I could nest components).

Forcing a full lifecycle solves these problems:
- It forces that first mount to use fresh data for all local state - including potentially derived state. 
- It allows concurrent rendering to function as intended. That initial state can be parallelized, and run concurrently. This does not happen to already mounted components.

### To Reproduce

To reproduce, create a route such as `/pages/[slug].js` and then create two pages to be handled within, with links to each other. Deliberately set initial state based on the server side props, and use that for rendering.

Observe that when changing pages, no concurrent rendering is used, and the data will be incorrect on the second page. Refreshing to force full lifecycle will show correct data.

```jsx
const Page = ({ pageFromStaticProps }) => {
  // contrived derived data example:
  const [page, setPage] = useState(pageFromStaticProps)
  // here we might use setPage to set a client only preview of edits from the CMS in useEffect, for example
  return <div>
    <Link href="/pages/about"><a>About</a></Link>
    <Link href="/pages/pool-party"><a>Pool Party</a></Link>
    {page.name} // observe, this will not change between page navigation
  <div>
}

const pages = {
  about: { name: "About" },
  "pool-party": { name: "Pool Party" }
}

getStaticProps...
getStaticPaths...
```

**Comment by ivan-kleshnin** [comment]
You can pass `key` in the `_app` if you want. 

But from the above code it seems you invent your own NextJS patterns. Like saving page props into the state to do... routing? 🙈 
I dunno, it's weird. If I saw such code at CR I'd just ask my team member to read the docs and try to rewrite it once again.

**Reply by CaptainN** [reply]
I did market is as "contrived derived data example:" - the problem I was having came from a more complex set of derived data based on data from getStaticProps - it really should set a key between pages, except when using shallow linking.
